src/web/mod.rs

```diff
@@ -310,33 +332,56 @@ fn match_version(conn: &Connection, name: &str, version: Option<&str>) -> MatchV
         .map(|v| if v == "newest" || v == "latest" { "*".to_owned() } else { v })
         .unwrap_or("*".to_string());
 
-    let versions = {
+    let (versions, corrected_name) = {
         let mut versions = Vec::new();
-        let rows = conn.query("SELECT versions FROM crates WHERE name = $1", &[&name]).unwrap();
-        if rows.len() == 0 {
-            return MatchVersion::None;
+        let mut rows = conn.query("SELECT versions FROM crates WHERE name = $1", &[&name]).unwrap();
+        let name_replace = if rows.len() == 0 {
+            // try looking up again with dashes and underscores replaced
+            if name.contains('-') {
+                Some(name.replace('-', "_"))
+            } else if name.contains('_') {
+                Some(name.replace('_', "-"))
+            } else {
+                return None;
+            }
+        } else {
+            // if we returned rows, then we don't need to check a replaced name
+            None
+        };
+
+        if let Some(new_name) = &name_replace {
+            rows = conn.query("SELECT versions FROM crates WHERE name = $1", &[&new_name]).unwrap();
+
+            if rows.len() == 0 {
+                // even with the new name, there was nothing, so bail
+                return None;
+            }
         }
+
         let versions_json: Json = rows.get(0).get(0);
         for version in versions_json.as_array().unwrap() {
             let version: String = version.as_string().unwrap().to_owned();
             versions.push(version);
         }
 
-        versions
+        (versions, name_replace)
     };
 
     // first check for exact match
     // we can't expect users to use semver in query
     for version in &versions {
         if version == &req_version {
-            return MatchVersion::Exact(version.clone());
+            return Some(MatchVersion {
+                corrected_name,
+                version: MatchSemver::Exact(version.clone()),
```

Comment:
Does crates.io allow packages with both underscores and hyphens? If so and we cache this page, someone could end up going to the wrong page if a crate is later published:

1. Someone tries to visit https://docs.rs/integer-set
2. They get served the page for https://docs.rs/integer_set
3. Cloudfront caches the page
4. A few months from now, someone publishes `integer-set`

Now all visits to `integer-set` will be cached and redirected to the wrong crate.
Instead this should be `MatchSemver::Semver` if `corrected_name` is not `None`
so it doesn't get cached and a proper redirect is issued.

Reply:
No, you can't publish `foo-bar` if `foo_bar` is already there.

Reply:
Ok, that makes sense. I think the redirect is still the right thing to do because users see the right URL and cloudfront has to cache less things.

Reply:
As currently implemented, the only handler that ever tries to process a corrected match is the `[/:crate, /:crate/:version]` handler which always redirects, so a `302 Found` redirect will be issued when going to `https://docs.rs/integer-set` pointing at `https://docs.rs/integer_set/0.1.0/integer_set` (or whatever version happens to be published).

All the other handlers (e.g. attempting to open `https://docs.rs/integer-set/0.1.0/integer_set` directly) will return a 404 as they do not allow the correction to occur.